Cache themes and retrieve from cache when available

banani720 opened the pull request:
#### Description

This pull request caches the theme data with an expiry date of 30 days to local storage. If a theme exists it is returned and no call to the CDN to get the theme are made

Closes #85 

#### What change does this PR introduce?

Please select the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (changes to docs/code comments)

#### What is the proposed approach?

To resolve this issue I simply added cache setting/getting functions in order to check for a cached theme before going to the CDN. If there is no cached theme or it is expired, a call to the CDN is made and the result is then cached for future use. I chose 30 days as an arbitrary number but we can expand or reduce that as needed

#### Checklist:

- [x] The commit message follows our adopted [guidelines](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Testing has been done for the change(s) added (for bug fixes/features)
- [x] Relevant comments/docs have been added/updated (for bug fixes/features)

tjtanjin reviewed (changes requested):
Hey @banani720! Thank you for the PR 😊 I'm viewing this from my mobile so I can't go through the changes thoroughly but a couple of things that came to my attention:
- The current implementation only caches the latest theme even if multiple themes are specified
- The cache key starts with "chatify" which does not match the package name 😅
- CSS styles text is not included in cache so the caching is only partial

If you have the time, those would be great to look at 😊

banani720 commented:
Hey @tjtanjin I'll get right on that, appreciate the feedback!

banani720 commented:
In the meantime, I've gone ahead and implemented a unique key for cached themes so that they don't overwrite and reworked `getCachedTheme` and `setCachedTheme` so that they handle undefined versions when setting cache keys. I've also gone ahead and redone the cache theme key to be `rcb-theme` in accordance with the naming standards I've seen elsewhere in the repo.

tjtanjin commented on src/services/ThemeService.tsx:
```diff
@@ -4,6 +4,42 @@ import { Theme } from "../types/Theme";
+const getCachedTheme = (id: string, version: string | undefined) => {
+	const cachedTheme = version ? localStorage.getItem(`${CACHE_KEY_PREFIX}_${version}_${id}`) 
+		: localStorage.getItem(`${CACHE_KEY_PREFIX}_${id}`);
+	if (!cachedTheme) return null;
```
Let's standardise to use Egyptian braces (no one-liner)

tjtanjin commented on src/services/ThemeService.tsx:
```diff
@@ -4,6 +4,42 @@ import { Theme } from "../types/Theme";
+	if (!cachedTheme) return null;
+
+	const themeData = JSON.parse(cachedTheme);
+	const now = new Date();
+	const expiryDate = new Date(themeData.expiry);
+
+	if (now > expiryDate) {
```
The current approach stores `expiry` in the cache data to represent the specific time where the cache is invalidated. This works, but let's say the developer makes a mistake of setting cache to 30 days and now wants to revert to 1 week. This will be an issue because the client side will continue to use the cache for up to 30 days.

A possible approach to circumvent this is to have `expiry` represent the time where the cache is being stored instead. Then when checking, take the `expiry` field and add on the valid period (e.g. 30 days or 1 week) before performing the check. This means that even if it was set to 30 days before, if the developer changes it to 1 week, the next time the client performs the check, it will utilise the new 1 week value.

tjtanjin commented on src/services/ThemeService.tsx:
```diff
@@ -71,5 +114,7 @@ export const processAndFetchThemeConfig = async (theme: Theme): Promise<{setting
 		console.error(`Failed to fetch styles.json from ${inlineStylesUrl}`);
 	}
 
+	setCachedTheme(`${CACHE_KEY_PREFIX}_${id}_${version}`, themeVersion, settings, inlineStyles);
```
CSS is not being cached here, but it should be cached.

tjtanjin reviewed (changes requested):
**Overall Comments:**
Thanks for the updates! It's better, but there're still some areas worth looking at. I list them in their own sections below.

**Suggested Improvement:**
The current approach doesn’t allow users to specify their cache duration, it’ll be great to give them the flexibility to do so.

Perhaps on top of having a default cache duration, we can include a fourth optional `cache_expiry` field inside the Theme type. This will allow users to specify the time in seconds before the theme cache expires.

 **Suggested Caching Logic (assumes above improvement is adopted):**
- If `cache_expiry` is specified, always use the value provided
- If `cache_expiry` is not specified, use the default expiry of 30 days
- If version is specified, then cache it directly (keep the current approach)
- If version is not specified, fetch the latest version from `meta.json` and retrieve the version to use (i.e. **do not** have a special cache for themes without version specified)

The above are just suggestions, feel free to share your thoughts as well 😊

tjtanjin commented:
Hey @banani720! Are you still working on this PR? If you require further clarifications feel free to ask 😊

tjtanjin commented:
Any updates @banani720? 🥹

tjtanjin commented:
Succeeded by #103
